src/Workspaces/Remote/Core/ServiceHubRemoteHostClient.cs

```diff
@@ -93,27 +93,47 @@ private void OnUnexpectedExceptionThrown(Exception unexpectedException)
+                try
+                {
+                    var hubClient = new HubClient("ManagedLanguage.IDE.RemoteHostClient");
+
+                    var remoteHostStream = await RequestServiceAsync(services, hubClient, WellKnownServiceHubService.RemoteHost, cancellationToken).ConfigureAwait(false);
 
-                var client = new ServiceHubRemoteHostClient(services, serviceBroker, serviceBrokerClient, hubClient, remoteHostStream, callbackDispatchers);
+                    client = new ServiceHubRemoteHostClient(services, serviceBroker, serviceBrokerClient, hubClient, remoteHostStream, callbackDispatchers);
 
-                var uiCultureLCID = CultureInfo.CurrentUICulture.LCID;
-                var cultureLCID = CultureInfo.CurrentCulture.LCID;
+                    var uiCultureLCID = CultureInfo.CurrentUICulture.LCID;
+                    var cultureLCID = CultureInfo.CurrentCulture.LCID;
 
-                // initialize the remote service
-                await client._endPoint.InvokeAsync<string>(
-                    nameof(IRemoteHostService.InitializeGlobalState),
-                    new object?[] { uiCultureLCID, cultureLCID },
-                    cancellationToken).ConfigureAwait(false);
+                    // initialize the remote service
+                    await client._endPoint.InvokeAsync<string>(
+                        nameof(IRemoteHostService.InitializeGlobalState),
+                        new object?[] { uiCultureLCID, cultureLCID },
+                        cancellationToken).ConfigureAwait(false);
 
-                await client.TryInvokeAsync<IRemoteAsynchronousOperationListenerService>(
-                    (service, cancellationToken) => service.EnableAsync(AsynchronousOperationListenerProvider.IsEnabled, listenerProvider.DiagnosticTokensEnabled, cancellationToken),
-                    cancellationToken).ConfigureAwait(false);
+                    await client.TryInvokeAsync<IRemoteAsynchronousOperationListenerService>(
+                        (service, cancellationToken) => service.EnableAsync(AsynchronousOperationListenerProvider.IsEnabled, listenerProvider.DiagnosticTokensEnabled, cancellationToken),
+                        cancellationToken).ConfigureAwait(false);
 
-                client.Started();
-                return client;
+                    client.Started();
+                    return client;
+                }
+                catch
+                {
+                    // Make sure ServiceBrokerClient is disposed on failure. At this point, it may or may not have been
+                    // wrapped in the ServiceHubRemoteHostClient instance, so handle both possibilities.
+                    if (client is not null)
+                    {
+                        client.Dispose();
+                    }
+                    else
+                    {
+                        serviceBrokerClient.Dispose();
+                    }
+
+                    throw;
```

Comment:
i so wish there was some sort of easier pattern in the language to handle this sort of thing.  any thoughts on if your ref-counted work could help here?

Reply:
The main way this is expected to be handled is by ensuring finalizers or weak references automatically clean up things created on paths that eventually lead to an exception. For this case, it sounds like an instance adds an event handler to a long-lived object using a strong reference instead of a weak reference, which in turn GC-roots the short-lived object created on an exception path.

Reply:
This strong reference in `ServiceBrokerClienc` is by design (and is the default event handler experience used in .NET and C#). Changing it now would be a breaking change.